burst-kotlin-plugin/src/main/kotlin/app/cash/burst/kotlin/InterceptorInjector.kt

```diff
@@ -0,0 +1,396 @@
+  fun adoptBeforeTest(function: IrSimpleFunction) {
+    function.annotations = function.annotations.filter {
+      it.type.classOrNull !in burstApis.beforeTestSymbols
+    }
+
+    beforeTestFunctionSymbols += function.symbol
+  }
+
+  /** Remove the `@AfterTest` annotation on [function]. */
+  fun adoptAfterTest(function: IrSimpleFunction) {
+    function.annotations = function.annotations.filter {
+      it.type.classOrNull !in burstApis.afterTestSymbols
+    }
+
+    afterTestFunctionSymbols += function.symbol
+  }
+
+  fun defineIntercept() {
+    check(interceptFunctionSymbol == null) { "already defined?!" }
+
+    // TODO: don't add 'implements TestInterceptor' if it already does.
+    originalParent.superTypes += burstApis.testInterceptor.defaultType
+
+    val function = originalParent.factory.buildFun {
+      initDefaults(originalParent)
+      name = Name.identifier("intercept")
+      returnType = pluginContext.irBuiltIns.unitType
+    }.apply {
+      modality = Modality.OPEN
+      overriddenSymbols += burstApis.testInterceptorIntercept
+      parameters += buildReceiverParameter {
+        type = originalParent.defaultType
+      }
+      addValueParameter {
+        initDefaults(originalParent)
+        name = Name.identifier("test")
+        type = burstApis.testInterceptorTest.defaultType
+      }
+    }
+
+    function.irFunctionBody(
+      context = pluginContext,
+      scopeOwnerSymbol = originalParent.symbol,
+    ) {
+      val packageNameLocal = irTemporary(
+        value = irCall(burstApis.testInterceptorTestPackageName.owner.getter!!).apply {
+          origin = IrStatementOrigin.Companion.GET_PROPERTY
+          dispatchReceiver = irGet(function.parameters[1])
+        },
+        nameHint = "packageName",
+      )
+
+      val classNameLocal = irTemporary(
+        value = irCall(burstApis.testInterceptorTestClassName.owner.getter!!).apply {
+          origin = IrStatementOrigin.Companion.GET_PROPERTY
+          dispatchReceiver = irGet(function.parameters[1])
+        },
+        nameHint = "className",
+      )
+
+      val functionNameLocal = irTemporary(
+        value = irCall(burstApis.testInterceptorTestFunctionName.owner.getter!!).apply {
+          origin = IrStatementOrigin.Companion.GET_PROPERTY
+          dispatchReceiver = irGet(function.parameters[1])
+        },
+        nameHint = "functionName",
+      )
+
+      var proceed: IrExpression = irBlock {
+        val failure = irTemporary(
+          value = irNull(),
+          irType = pluginContext.irBuiltIns.throwableType.makeNullable(),
+          nameHint = "failure",
+          isMutable = true,
+        )
+
+        +irAccumulateFailure(
+          burstApis,
+          failure,
+          irBlock {
+            // Call each function annotated `@BeforeTest` in reverse alphabetical order.
+            for (functionSymbol in beforeTestFunctionSymbols.sortedBy { it.owner.name }.reversed()) {
```

Comment:
This ordering (and the two below) feels surprising. I would expect declaration order for all three.